### PR TITLE
Type state verifier, refactored verifiable timestamp, filtering for block commitments

### PR DESCRIPTION
**Pull request by sgreenbury**
This PR includes:
- Type state pattern for `Verifier` with `FullClient` and `LightClient` (#101)
- Refactor of `VerifiableTimestamp` (#102)
- Filtering for block commitments and revised commitment traits with generics (#104)

### CODE REVIEW

File: trustchain-http/src/resolver.rs [hunk 1]
```diff
@@ -221,17 +222,20 @@ mod tests {
...
         // Test for case where incorrect root_event_time for the root of the given DID, expected to
         // return Ok but with a JSON containing the wrapped Trustchain error.
-        let incorrect_root_event_time = "001234500";
+        let incorrect_root_event_time = 1234500;
         let uri_incorrect_root = format!(
             "/did/chain/did:ion:test:EiAtHHKFJWAk5AsM3tgCut3OiBY4ekHTf66AAjoysXL65Q?root_event_time={incorrect_root_event_time}"
         )
         .to_string();
         let response = client.get(&uri_incorrect_root).send().await;
+        // TODO: fix test to handle error that is now returned during verififcation
```
Comment (sgreenbury) [inline trustchain-http/src/resolver.rs:231]: To fix.

File: trustchain-core/src/verifier.rs [hunk 2]
```diff
@@ -263,10 +205,16 @@ pub trait Verifier<T: Sync + Send + DIDResolver> {
...
+        // Verify that the DID content and the timestamp share a common commitment target.
+        verifiable_timestamp.verify(&verifiable_timestamp.timestamp_commitment().hash()?)?;
+
+        // Validate the PoW on the common target hash.
+        self.validate_pow_hash(&verifiable_timestamp.timestamp_commitment().hash()?)?;
+
+        // At this point we know that the same, valid PoW commits to both the timestamp
         // in verifiable_timestamp and the data (keys & endpoints) in the root DID Document.
         // It only remains to check that the verified timestamp matches the expected root timestamp.
         if !verifiable_timestamp.timestamp().eq(&root_timestamp) {
```
Comment (sgreenbury) [inline trustchain-core/src/verifier.rs:220]: Confirm whether this check is no longer required.

File: trustchain-http/src/resolver.rs (hunk 1 above)
Comment (sgreenbury) [reply trustchain-http/src/resolver.rs]: Resolved by #112

File: trustchain-core/src/verifier.rs (hunk 2 above)
Comment (sgreenbury) [reply trustchain-core/src/verifier.rs:225]: Resolved in #112

**Review by thobson88 — APPROVED** [review]
LGTM

**Pull request merged**